tick/eval/types.go

```diff
@@ -0,0 +1,189 @@
+package eval
+
+import (
+	"reflect"
+	"regexp"
+
+	"github.com/influxdata/kapacitor/tick"
+)
+
+type ValueType uint8
+
+const (
+	InvalidType ValueType = iota << 1
+	TFloat64
+	TInt64
+	TString
+	TBool
+	TRegex
+)
+
+const TNumeric = TFloat64 & TInt64
+
+func isNumericValueType(valueType ValueType) bool {
+	return (valueType|TInt64 == TInt64) || (valueType|TFloat64 == TFloat64)
+}
+
+var valueTypeToName = map[ValueType]string{
+	TFloat64:    "float64",
+	TInt64:      "int64",
+	TString:     "string",
+	TBool:       "boolean",
+	TRegex:      "regex",
+	InvalidType: "invalid type",
+}
+
+func valueTypeOf(t reflect.Type) ValueType {
+	switch t.Kind() {
+	case reflect.Float64:
+		return TFloat64
+	case reflect.Int64:
+		return TInt64
+	case reflect.String:
+		return TString
+	case reflect.Bool:
+		return TBool
+	default:
+		// This is not primitive type.. so it must be regex
+		if t == reflect.TypeOf((*regexp.Regexp)(nil)) {
+			return TRegex
+		}
+		return InvalidType
+	}
+}
+
+// getCostantNodeType - Given a tick.Node we want to know it's return type
+// this method does exactly this, few examples:
+// *) StringNode -> TString
+// *) UnaryNode -> we base the type by the node type
+func getConstantNodeType(n tick.Node) ValueType {
+	switch node := n.(type) {
+	case *tick.NumberNode:
+		if node.IsInt {
+			return TInt64
+		}
+
+		if node.IsFloat {
+			return TFloat64
+		}
+	case *tick.StringNode:
+		return TString
+	case *tick.BoolNode:
+		return TBool
+	case *tick.RegexNode:
+		return TRegex
+
+	case *tick.UnaryNode:
+		// If this is comparison operator we know for sure the output must be boolean
+		if node.Operator == tick.TokenNot {
+			return TBool
+		}
+
+		// if this is math result it can be Int64/Float64, the type decision will be upon
+		// getConstantNodeType to choose or if it's dynamic node the EvalUnaryNode will choose.
+		if node.Operator == tick.TokenMinus {
+			nodeType := getConstantNodeType(node.Node)
+			if nodeType == InvalidType {
+				return TNumeric
+			}
+
+			return nodeType
+		}
+
+	case *tick.BinaryNode:
+		if tick.IsCompOperator(node.Operator) {
+			return TBool
+		}
+
+		if tick.IsMathOperator(node.Operator) {
+			// Numeric result shall return, we must be more specific
+			leftType := getConstantNodeType(node.Left)
+
+			// quick exit here..
+			if leftType == TFloat64 {
+				return TFloat64
+			}
+
+			rightType := getConstantNodeType(node.Right)
+			if rightType == TFloat64 {
+				return TFloat64
+			}
+
+			if leftType == TInt64 && rightType == TInt64 {
+				return TInt64
+			}
+
+			// This is math operator but now float or int,
+			// This is invalid expression which will get an error while evaluating
+			// We can return an error right here if we want
+			return TNumeric
+		}
+	}
+
+	return InvalidType
+}
+
+func isDynamicNode(n tick.Node) bool {
```

Comment:
I think it is fine how it is. It is clear that is determining whether a give tick.Node has dynamic return type. Was there an issue with it that I am missing?